### DISCUSSION TITLE
Guidance on caching data during CompilerPass

### DISCUSSION THREAD

**Discussion by tacman** [symfony/symfony#52760] [Q&A] [answered]
I'm working on a menu bundle and would like to make sure the user has permission before adding an element to the menu.  That is, I'd like addMenuItem to check for the isGranted so that I don't need this:

```php
if ($this->isGranted('ROLE_ADMIN')) {
    $this->addMenuItem('app_import_data');
}
```

To do that, I scan the controllers (using ReflectionClass) and look for the #[IsGranted()] attribute.  

Obviously, I want to only scan the directories during the compiler pass, and store the IsGranted value with the route, so

```php
$routePermissions  = ['app_import_data' => 'ROLE_ADMIN'];
```

And passing that data to my bundle, so that addMenuItem can do the check.

What I'm missing is how to save that data so that the reflection is only done once.  I know I need to create that data, and then store like other cached data is stored (translations, classmap, symfony routes, etc.).  

I assume I want to use https://symfony.com/doc/current/components/config/caching.html, but I'm still having a bit of a hard time wrapping my head around it.  Are there any examples of a relatively simple bundle that does this?

Thanks.

**Comment by ghost** [comment]
`file_put_contents($data, $kernelCacheDir)`

**Reply by tacman** [reply]
and then how / when is that read?  Or does that happen automatically?  if so, that was the piece I was missing.

**Reply by ghost** [reply]
what is read? The cache created via file_put_contents? It's your bundle so it's up to you when you read the cache you created before in with the Compiler Pass.

**Reply by tacman** [reply]
Ah, I see.  In the compilerPass, write the data.  Then in the build() call of the bundle, read the data from that file and inject it into the service that needs it.  I'll give that a try.  Thanks!!

**Reply by tacman** [reply]
Thanks for your help with this.  I'm still debugging my code, but it seems like it's happening in the wrong order, that process() happens after loadExtension().

Perhaps I shouldn't be reading the cached file in loadExtension(), but somewhere else?  

In process() (the compiler pass), I write my data, if loadExtension is called first, is there somewhere else I can inject the cached contents into a service?

**Reply by ghost** [reply, marked as answer]
Yes, extensions run before compiler passes. Just read the cache when the app is running i.e. when you decide whether to render a menu or not.

**Reply by tacman** [reply]
So inject the filename into the bundle service, then read it when it's called.  Makes sense!
